route53_health_check: bugfix: port comparison

napkindrawing opened the pull request:
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
route53_health_check

##### ANSIBLE VERSION
```
ansible 2.2.1.0
  config file =
  configured module search path = Default w/o overrides
```

##### CONFIGURATION
(none)

##### OS / ENVIRONMENT
N/A

##### SUMMARY

The `route53_health_check` module is mistakenly comparing `unicode` and `int` values for `port` and in so doing always thinks an update is required when one may not be.

When an existing health check configuration is fetched and stored locally for comparison, the `port` is represented as `unicode`, but the local ansible configuration represents `port` as `int`, which results in `!=` always returning `True` even when one is `u'80'` and the other is `80`.

This fix simply checks to see if the existing/wanted comparison is comparing `unicode` and `int` values, and, if so, uses `str(...)` for the value of the given `int` value instead.

##### STEPS TO REPRODUCE

Create a health check with the `route53_health_check` module.

Make sure you have the update fix from https://github.com/ansible/ansible/pull/25976

Don't change anything in your definition, and re-run your original `route53_health_check` and you'll see the health check be needlessly updated.

##### EXPECTED RESULTS

Updates when changes are present, no updates when no changes are present.

##### ACTUAL RESULTS

Updates always applied and executed even when no configuration has changed.

napkindrawing committed:
route53_health_check: bugfix: port comparison

When an existing health check configuration is fetched and stored locally for comparison, the `port` is represented as `unicode`, but the local ansible configuration represents `port` as `int`, which results in `!=` always returning `True` even when one is `u'80'` and the other is `80`.

This simply checks to see if the existing/wanted comparison is comparing `unicode` and `int` values, and, if so, uses `str(...)` for the value of the given `int` value instead.

ryansb commented on lib/ansible/modules/cloud/amazon/route53_health_check.py:
```diff
@@ -185,8 +185,16 @@ def health_check_diff(a, b):
+        b_val = b.get(key)
+        a_type = type(a_val)
+        b_type = type(b_val)
+        if (a_type is int and b_type is unicode):
```
Instead of unicode, it's safest to compare against text_type since Py2 and Py3 have different types. 

Alternatively, it seems it'd be sufficient to do: `if int(a.get(key)) != int(b.get(key))` since running `int()` on an integer is a no-op.

napkindrawing replied:
Ah, thanks Ryan, I'm not primarily a Python person so I was wondering about that.

I just updated it to run the values through int() but are you saying don't even compare to the type `unicode` in there? Should it be something like this instead?

`if (a_type is int and b_type is not int):` ?

napkindrawing committed:
route53_health_check: fix unicode/int compare

napkindrawing committed:
route53_health_check: remove references to unicode

For Py2/Py3 compatibility, just see if we need to convert for int/not-int comparisons

napkindrawing reviewed (commented):
Hi Ryan,
I've updated the branch with an explicit conversion of non-int to int as you'd mentioned.

s-hertel commented:
Hi @napkindrawing I don't run into the always changed error on the fix #25706, but I'm wondering if you would be willing to check me on that.

s-hertel commented:
Closing because I think this was fixed in e49f15d6e4d735ba6f58e3e15b2353e4781c7e6b.
